src/login/service.ts

```diff
@@ -228,21 +230,11 @@ async function challengeUsernamePassword(session: LoginSession, username: string
     );
   }
 
-  if (!await services.user.validatePassword(user, password)) {
-    throw new A12nLoginChallengeError(
-      session,
-      'Incorrect username or password',
-      'username-password',
-      true,
-    );
-  }
-
   session.principalId = user.id;
   session.passwordValid = true;
   session.dirty = true;
```

Comment:
For context, this is a new service I've been working on, and the goal is to eventually handle all login flows in a consistent way including all the special cases like TOTP, WebAuthN and other MFA flows.

The issue here is that this service is very imperative, so it takes one step at the time to 'build up' the entire login session and ask for all the credentials/factors.

Moving the login check later in this function means that you set 'passwordValid' to true before the actual check is done. This information gets saved in the 'login session', so even if the password was wrong, you end up with a session that no longer requires a password check. See line 240-242 in the old diff and 233-235 in the new one.

This is actually the flow we use for sprout now. Once this is fixed I think this is good to go!

Reply:
@evert Updated: 3a52f2fc2409d9c0bd0cd38bfa216db619dbb259

For context, I moved this logic because the `validateUserCredentials` method calls `await loginActivityService.resetFailedLoginAttempts(user);`. My intention was to delay the execution of `validateUserCredentials` method as much as possible so that `resetFailedLoginAttempts` would run only when the login actually succeeds.

However, in this case, that approach is not feasible because as you mentioned, we need to execute the `validateUserCredentials` method before setting `session.passwordValid = true;`. Thank you for pointing that out!

I also considered running the `resetFailedLoginAttempts` method imperatively in each controller/service. However, I believe the current approach is fine since we call the `validateUserCredentials` method just before logging the `loginSuccess` event in all controllers/services, except for this one. And it makes sense to run the `resetFailedLoginAttempts` method after checking all relevant aspects in the `validateUserCredentials` method,  such as `isAccountLocked`, `validatePassword` or `reachedMaxAttempts`, even while other checks for login are still ongoing.

Reply:
This works for me! Thank you. Login Service is still a work in progress so I probably need to spend some more time making it's logic a bit more clear